Add RNA-Seq clustering heatmap example notebook

cbethell opened the pull request:
## Purpose:
What issue(s) does your PR address?

This PR addresses the issue creating a clustering example with RNA-seq data (#127).

As noted in one of my [comments on that issue](https://github.com/AlexsLemonade/refinebio-examples/issues/127#issuecomment-656948685), I followed a plan in the order as follows:

1. Copy the microarray clustering notebook and rename the module and notebook to include the word `heatmap`.
2. Find a suitable RNA-Seq dataset (with non-QN'ed data) and placing it into the `data`.
3. Create the DESeqDataSet.
4. Add a step to define a minimum counts cutoff. 
5. Perform the DESeq2 normalization and use `assay()` to further analyze the now normalized data.
6. Keep the section that chooses genes of interest using variances.
7. Use the `pheatmap` package to create a basic clustered heatmap and add an additional heatmap to show that the clustering can be turned off.

## Questions for reviewers

- Does the dataset seem like an appropriate dataset to use for this example (specifically worried about the heatmap)?
- Are there any suggestions for a less precarious way of defining the `Groups` variable for the heatmap annotation? (Still looking into this myself)
- Are the comments clear/suffice?

## Pull Request Check List:
* [x] Run a linter
* [x] Set the seed (if applicable)
* [x] Comments and/or documentation up to date
* [x] Double check your paths
* [x] Check headline numbering
* [x] Spell check any Rmd file or md file
* [x] Restart R and run all notebooks fresh and save

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
````diff
@@ -0,0 +1,220 @@
+you will see later using a function from the `DESeq2` package. This chunk of
+code will turn the data.frame (or matrix) into a `DESeqDataSet` object.
+
+```{r Create a DESeq dataset}
````
This likely going to be a thing we do many times in the whole RNA-seq section's many modules so we may want to have it pretty well honed down. Can also add an explanation about `DESeqDataSet`'s and etc in the RNA-seq header section (#125) which we can link to here.

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
```diff
@@ -0,0 +1,220 @@
+
+# Define a `Groups` column variable
+col_df <- col_df %>%
+  dplyr::mutate(Groups = rep(c("SP", "MP"), # Replace with group names relevant to your data
```
In other words, can you explain why you used this `colData(dds)` strategy you are using here? It isn't clear to me, but this could because I haven't tried this myself.

cbethell replied:
I will give your first suggestion a try and let you know from there. I did try something similar before and that did not work, but I believe it was due to the way I generated the `Groups` column in that case.

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
```diff
@@ -0,0 +1,220 @@
+
+The data that we downloaded from refine.bio for this analysis is a dataset, with
+a total of four samples, containing RNA-sequencing results for mouse primary
+natural killer (NK) cells and NK cells treated with DKK2 for 24 hours.
```
Ignore this comment until we get a new, bigger dataset

cbethell commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
````diff
@@ -0,0 +1,220 @@
+```
+
+## 2) Import and set up data
+Data downloaded from refine.bio include a metadata tab separated values ("tsv")
````
Good point!

cbethell commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
```diff
@@ -0,0 +1,220 @@
+# Calculate the variance for each gene
+variances <- apply(df, 1, var)
+
+# Determine summary statistics for gene variances
```
Sounds good and no worries!

cbethell commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
````diff
@@ -0,0 +1,220 @@
+df.by.var <- df[which(variances > sum.stats.var[5]), ]
+```
+
+## 4) Create a DESeq dataset
````
Another good point 👍

cbethell commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
````diff
@@ -0,0 +1,220 @@
+
+```{r Choose genes}
+# Calculate the variance for each gene
+variances <- apply(df, 1, var)
````
I think it's a good idea 👍 (we do use the `apply` function elsewhere)

cbethell replied:
This is the only comment I have yet to address because I have not yet thought of an appropriate tidyverse strategy to replace what the apply function is doing here.

jashapiro replied:
Apply is fine here. This is essentially a matrix, and you are doing a row-wise calculation, which tidyverse is less adept at. 
A suggestion you could make is to use `matrixStats::rowVars()`, which is much faster than `apply()`, but I think apply is fine.

cbethell commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
````diff
@@ -0,0 +1,220 @@
+
+```{r Define minimum counts cutoff}
+# Define a minimum counts cutoff
+genes_to_keep <- rowSums(counts(dds)) >= 10
````
Sounds good to me 👍

cbethell commented:
> I think this is great start to get the ball rolling! Honestly this PR brought up a lot of things for us to think about. With a lot of them related to #125 - I'm going to put these notes down on that issue.
> 
> I have a lot of comments, but tbh, a lot of them are to 2018 Candace's choices (I probably should have made it more clear to you that you are free to burn this whole notebook down and start fresh if you so want). But while we are revamping these notebooks, we should probably take the opportunity to change anything we don't think is ideal. (No time like the present I suppose).
> 
> But now that we have the ball rolling, I think the bigger change we may want to make is to use a dataset with more samples since a clustering of n = 4 is not super helpful (Thanks a LOT 2018 Candace for the bad example).
> 
> Here's what I think the plan should be moving forward:
> 
> 1. Find a bigger RNA-seq dataset with metadata labels we can use - feel free to post your ideas on the issue #127 for feedback.
> 2. Make this notebook fit that dataset. I would try it out a "quick and dirty" clustering on it before revamping the rest so we don't get to attached - feel free to post the plot you get. We won't want a dataset that clusters to nonsensically or too perfectly - something that's a good, realistic example but has sensible enough metadata labels we can use for trying to ground this in some semblance of truth.
> 3. Then we can go through addressing comments listed here (except for the RNA-seq header links parts).
> 4. For the time being, you can ignore the items where I said we will need to link it to the RNA-seq header section (#125) and just put a `TODO: link to RNA-seq header` I'll write the things we've found here on #125 and when that gets addressed we can come back put the links.
> 
> Let me know if what I'm saying here makes sense, feel free to send me your ideas.

It all makes sense, I'll implement the changes/steps you suggested here update the PR along the way. I will also take you up on that suggestion of posting on the issue for feedback on the dataset and plot 👍 Thank you!

jashapiro commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
```diff
@@ -0,0 +1,296 @@
+## TODO: Link to explanation about `DESeqDataSet` in the RNA-seq header section
+df_by_var <- df_by_var %>%
+  # Mutate numeric variables to be integers
+  dplyr::mutate_if(is.numeric, as.integer)
```
`as.integer()` always rounds down, so it should be avoided.
```suggestion
  dplyr::mutate_if(is.numeric, round)
```

cbethell replied:
Gotcha 👍

cbethell commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
```diff
@@ -0,0 +1,296 @@
+## TODO: Add a note/link to section in documentation on why we use
+# `normTransform` vs `vst` and `rlog`
+# Let's normalize the data for the collapsed counts first 
+ntd_collapsed <- normTransform(dds_collapsed)
```
I left this TO DO because I am not sure myself. My decision came from the original issue #127 where you mentioned `DESeq2` normalization, so I figured that the `normTransform`, being a function from the `DESeq2` package, was the best to implement here and it was also implemented in the vignette you put on the original issue.

cansavvy replied:
When you have time (so maybe not before we get training stuff wrapped up), I would look at these docs for: 
- [`normTransform`](https://www.rdocumentation.org/packages/DESeq2/versions/1.12.3/topics/normTransform) 
and see if you can decipher how this is different than 
- [`rlog`](https://www.rdocumentation.org/packages/DESeq2/versions/1.12.3/topics/rlog) or 
- [`vst`](https://www.rdocumentation.org/packages/DESeq2/versions/1.12.3/topics/varianceStabilizingTransformation)

We can chat about it sometime. The paper might also be helpful to you: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4302049/

cbethell replied:
Gotcha! :thankyou:

cbethell commented:
Thank you for the feedback @cansavvy! I _believe_ I rearranged the notebook as you described in your [comment above](https://github.com/AlexsLemonade/refinebio-examples/pull/140#pullrequestreview-452804156) and addressed your smaller review comments.

Please let me know if I misinterpreted any of your comments! Also, the comment that I added at the end on the results represented on the heatmap could probably use more detail, but I wanted to ensure that this is the heatmap that we want (however, feel free to comment on my lack of detail there if you agree and possibly what you would like to see added there).

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
```diff
@@ -0,0 +1,338 @@
+This chunk of code will read the both tsv files and add them as data.frame
+objects to your environment.
+
+The data that we downloaded from refine.bio for this analysis is a dataset, with
```
```suggestion
The data that we downloaded from refine.bio for this analysis has
```

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
```diff
@@ -0,0 +1,338 @@
+objects to your environment.
+
+The data that we downloaded from refine.bio for this analysis is a dataset, with
+a total of 24 samples (obtained from 6 mice with 4 technical replicates for each
```
```suggestion
24 samples (obtained from 6 mice with 4 technical replicates for each
```

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
```diff
@@ -0,0 +1,338 @@
+orthotopic xenograft mice with and without digoxin treatment.
+
+You can download this dataset by going to the following link and clicking 
+`Download Now`: https://www.refine.bio/experiments/SRP150101/transcriptome-analysis-of-group-3-and-4-medulloblastoma-orthotopic-xenograft-mice-with-digoxin-treatment
```
This part is actually not necessary here, because it will be covered when we add in the getting started section for this module (being workshopped over on #122 ).

cansavvy replied:
But you can keep it here for now I guess.

cbethell replied:
In your [previous review comment](https://github.com/AlexsLemonade/refinebio-examples/pull/140#discussion_r456024079), you mentioned that there is a likelihood of someone skipping over this in the getting started section so it is good to have here as well.

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
```diff
@@ -0,0 +1,338 @@
+)) %>%
+  dplyr::select(refinebio_accession_code, refinebio_title, refinebio_treatment) %>%
+  replace(is.na(.), "None") %>%
+  dplyr::mutate(
```
Can we put these `mutate` steps in their own chunk? This way the user has a chance to see what `metadata` looked like before. Would also be good to add a chunk in between where you print out a preview of `metadata`.

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
````diff
@@ -0,0 +1,338 @@
+```
+
+## How are our technical replicates clustering together?
+
````
This is great to start out with the question like we discussed. Can you try to add some words here to explain why we are doing this and give a brief summary of what we will be doing?

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
````diff
@@ -0,0 +1,338 @@
+
+```{r}
+# Let's prepare the metadata for the uncollapsed `DESeqData` set object
+metadata_uncollapsed <- metadata %>%
````
This metadata will only be used for annotation correct? 
If yes, can we instead call this something like "annotation_df" or something like that? 

This is in part because I'd like to move away from using `collapsed` `uncollapsed` terminology and instead try to use terms that are more reflective of the samples and the question at hand. I think this will be helpful for our users to keep focused on the question/purpose of what we are doing.

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
````diff
@@ -0,0 +1,338 @@
+```{r}
+# Let's prepare the metadata for the uncollapsed `DESeqData` set object
+metadata_uncollapsed <- metadata %>%
+  dplyr::select(-c("sample_info_summary", "refinebio_title")) %>%
````
Can you add a comment to explain what you are doing here? aka why are we dropping these two columns?

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
```diff
@@ -0,0 +1,338 @@
+# Let's prepare the metadata for the uncollapsed `DESeqData` set object
+metadata_uncollapsed <- metadata %>%
+  dplyr::select(-c("sample_info_summary", "refinebio_title")) %>%
+  tibble::column_to_rownames("refinebio_accession_code")
```
Also comment here for why we are storing these as rownames (cuz pheatmap I assume).

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
````diff
@@ -0,0 +1,338 @@
+```
+
+### 6) Define a minimum counts cutoff
+We are going to do some pre-filtering to keep only rows with 10 or more reads
````
Can you add a sentence or two about why we are doing this? Users may not know why.

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
````diff
@@ -0,0 +1,338 @@
+```{r Define minimum counts cutoff}
+# Let's take a look at the distribution of counts first
+
+# Sum up the counts for each row
````
We might be a little more comfortable with rows = genes and columns = samples. But our users are probably not as comfortable with this. Can you bring these steps back to what they mean in terms of biology?

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
```diff
@@ -0,0 +1,338 @@
+
+## How are our biological replicates clustering together?
+
+### 4) Create a DESeqDataset
```
We may want to get rid of the numbering? What do you think?

cbethell replied:
I had the same though as it could be a bit confusing, although I also do think that it can be helpful in allowing users to keep track of the steps sequentially if they go back to the notebook at some point and want to use it to answer just one of the questions.

cansavvy replied:
Can we go back and make it 4a and 4b or is that more confusing?

cansavvy commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
```diff
@@ -0,0 +1,338 @@
+## How are our biological replicates clustering together?
+
+### 4) Create a DESeqDataset
+A `DESeqDataSet` object is required to carry out the normalization step that
```
I'm debating about whether we need to explain this or not again in the same notebook. Unsure. My default thought is to keep it.

cbethell commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
````diff
@@ -0,0 +1,387 @@
+```{r Perform DESeq2 normalization}
+# Normalize the data in the `DESeqDataSet` object using the
+# `normTransform` function from the `DESEq2` R package
+## TODO: Add a note/link to section in documentation on why we use
````
Gotcha.

cansavvy commented:
> we used much of what we have here (e.g., the dataset and collapsing steps) as a DGE example?

I like this idea. So if we move this to DGE (probably delete the cluster and heatmap steps). Find a simpler/lower complexity dataset for this clustering example so we can remove the collapse/uncollapse steps from here?

cbethell commented on 03-rnaseq/clustering/clustering_heatmap_example.Rmd:
````diff
@@ -0,0 +1,435 @@
+# only rows that have counts above the cutoff
+genes_to_keep_all_reps <- rowSums(counts(dds_all_reps)) >= 10
+dds_all_reps <- dds_all_reps[genes_to_keep_all_reps, ]
+```
````
This was included per @cansavvy's review suggestion [here](https://github.com/AlexsLemonade/refinebio-examples/pull/140#discussion_r456042011), but a summary of the reasoning here is to plot the distribution of counts before defining a cutoff.

cbethell commented:
> > we used much of what we have here (e.g., the dataset and collapsing steps) as a DGE example?
> 
> I like this idea. So if we move this to DGE (probably delete the cluster and heatmap steps). Find a simpler/lower complexity dataset for this clustering example so we can remove the collapse/uncollapse steps from here?

I also have the same question here @jaclyn-taroni. I started to look for a simpler dataset and can post one or two once I find them to get feedback on whether it's a dataset that would be better suited for the notebook.

jaclyn-taroni commented:
Yes I agree with:

> Find a simpler/lower complexity dataset for this clustering example so we can remove the collapse/uncollapse steps from here

Keep an eye on how much metadata smoothing is required, too, as you're picking a dataset. My understanding is that a lot of the metadata "stuff" that has to happen here is also because of the replicates.

cbethell commented:
Any thoughts on whether or not the following dataset would be better suited for this notebook?

[A role for p53 in the adaptation to glutamine starvation](https://www.refine.bio/experiments/SRP151009/a-role-for-p53-in-the-adaptation-to-glutamine-starvation-through-the-expression-of-slc1a3#samples)

<img width="798" alt="Screen Shot 2020-08-04 at 5 24 22 PM" src="https://user-images.githubusercontent.com/43576623/89346829-b0918300-d677-11ea-9c3d-e6c79b7e96e9.png">

I am still in search for a dataset with a larger number of individual samples and lower complexity, but wanted to put this here on the ticket in the meantime in case someone thinks we should go ahead with this one.

jaclyn-taroni commented:
Now that the plan for using this notebook to get started on a DGE example is detailed on #137 and we have a plan to move forward with clustering specifically (https://github.com/AlexsLemonade/refinebio-examples/issues/127#issuecomment-669347715), I am going to close this pull request, retain the `cbethell/add-clustering-rna-seq-example` and disconnect this PR from #127.
